src/rpc/tensorpipe/tp_communicator.cc

```diff
@@ -0,0 +1,162 @@
+  std::string* zerocopy_blob_ptr = &tp_msg.metadata;
+  StreamWithBuffer zc_write_strm(zerocopy_blob_ptr, true);
+  zc_write_strm.Write(msg);
+  int32_t nonempty_ndarray_count = zc_write_strm.buffer_list().size();
+  zerocopy_blob_ptr->append(reinterpret_cast<char*>(&nonempty_ndarray_count),
+                            sizeof(int32_t));
+  tp_msg.tensors.resize(nonempty_ndarray_count);
+  // Hold the NDArray that ensure it's valid until write operation completes
+  auto ndarray_holder = std::make_shared<std::vector<NDArray>>();
+  ndarray_holder->resize(nonempty_ndarray_count);
+  auto& buffer_list = zc_write_strm.buffer_list();
+  for (int i = 0; i < buffer_list.size(); i++) {
+    auto& ptr = buffer_list[i];
+    (*ndarray_holder.get())[i] = ptr.tensor;
+    tensorpipe::CpuBuffer cpu_buffer;
+    cpu_buffer.ptr = ptr.data;
+    tp_msg.tensors[i].buffer = cpu_buffer;
+    tp_msg.tensors[i].length = ptr.size;
+    if (ptr.size == 0) {
+      LOG(FATAL) << "Cannot send a empty NDArray.";
+    }
+  }
+  pipe->write(tp_msg, [ndarray_holder](const tensorpipe::Error& error) {
+    if (error) {
+      LOG(FATAL) << error.what();
+    }
+  });
+}
+
+void TPSender::Finalize() {}
+void TPReceiver::Finalize() {}
+
+bool TPReceiver::Wait(const std::string& addr, int num_sender) {
+  std::string addr_str(addr);
+  listener = context->listen({addr});
+  for (int i = 0; i < num_sender; i++) {
+    std::promise<std::shared_ptr<Pipe>> pipeProm;
+    listener->accept([&](const Error& error, std::shared_ptr<Pipe> pipe) {
+      if (error) {
+        LOG(WARNING) << error.what();
+      }
+      pipeProm.set_value(std::move(pipe));
+    });
+    std::shared_ptr<Pipe> pipe = pipeProm.get_future().get();
+    std::promise<bool> checkConnect;
+    pipe->readDescriptor(
+      [pipe, &checkConnect](const Error& error, Descriptor descriptor) {
+        Allocation allocation;
+        checkConnect.set_value(descriptor.metadata == "dglconnect");
+        pipe->read(allocation, [](const Error& error) {});
+      });
+    CHECK(checkConnect.get_future().get()) << "Invalid connect message.";
+    pipes_[i] = pipe;
+    ReceiveFromPipe(pipe, queue_);
+  }
+  return true;
+}
+
+void TPReceiver::ReceiveFromPipe(std::shared_ptr<Pipe> pipe,
+                                 std::shared_ptr<RPCMessageQueue> queue) {
+  pipe->readDescriptor([pipe, queue = std::move(queue)](const Error& error,
+                                                        Descriptor descriptor) {
+    if (error) {
+      // Error may happen when the pipe is closed
+      return;
+    }
+    Allocation allocation;
+    CHECK_EQ(descriptor.payloads.size(), 0) << "Invalid DGL RPC Message";
+
+    int tensorsize = descriptor.tensors.size();
+    if (tensorsize > 0) {
+      allocation.tensors.resize(tensorsize);
+      for (int i = 0; i < descriptor.tensors.size(); i++) {
+        tensorpipe::CpuBuffer cpu_buffer;
+        cpu_buffer.ptr = new char[descriptor.tensors[i].length];
```

Comment:
This will be released at https://github.com/dmlc/dgl/blob/c8b18b793c5e8c1d3cd81060b6f6c1540a52c127/src/graph/serialize/zerocopy_serializer.cc#L24. Basically the CpuBuffer here will be used to construct a NDArray, which owns the underlying pointer. Therefore it will free the memory once destructed.